tensorflow/core/kernels/substr_op.cc

```diff
@@ -0,0 +1,220 @@
+#include "tensorflow/core/framework/types.h"
+#include "tensorflow/core/kernels/bounds_check.h"
+#include "tensorflow/core/lib/core/errors.h"
+#include "tensorflow/core/util/bcast.h"
+
+namespace tensorflow {
+
+// Position/length can be 32 or 64-bit integers
+template <typename T>
+class SubstrOp : public OpKernel {
+  public:
+    using OpKernel::OpKernel;
+
+    void Compute(OpKernelContext* context) override {
+      // Get inputs
+      const Tensor& input_tensor = context->input(0);
+      const Tensor& pos_tensor = context->input(1);
+      const Tensor& len_tensor = context->input(2);
+      const TensorShape input_shape = input_tensor.shape();
+      const TensorShape pos_shape = pos_tensor.shape();
+      const TensorShape len_shape = len_tensor.shape();
+
+      bool is_scalar = TensorShapeUtils::IsScalar(pos_shape);
+      
+      if (is_scalar || input_shape == pos_shape) {
+        // pos/len are either scalar or match the shape of input_tensor
+        // Do not need to do broadcasting
+
+        // Reshape input 
+        auto input = input_tensor.flat<string>();
+        // Allocate output
+        Tensor* output_tensor = nullptr;
+        OP_REQUIRES_OK(context,
+                       context->allocate_output("output", input_tensor.shape(),
+                                                &output_tensor));
+        auto output = output_tensor->flat<string>();
+        if (is_scalar) {
+          // Perform Op with scalar pos/len
+          const T pos = tensorflow::internal::SubtleMustCopy(pos_tensor.scalar<T>()());
+          const T len = tensorflow::internal::SubtleMustCopy(len_tensor.scalar<T>()());
+          for (size_t i = 0; i < input_tensor.NumElements(); ++i) {
+            // Make sure pos won't cause a runtime error
+            OP_REQUIRES(context, pos >= 0 && pos < input(i).size(),
+                        errors::InvalidArgument("pos ", pos, 
+                                                " out of range for string b'", 
+                                                input(i), "' at index ", i));
+            output(i) = input(i).substr(pos, len);
+          }
+        } else {
+          // Perform Op element-wise with tensor pos/len
+          auto pos_flat = pos_tensor.flat<T>();
+          auto len_flat = len_tensor.flat<T>();
+          // Use SubtleMustCopy on pos/len to prevent async attacks
+          const size_t num_elements = pos_tensor.NumElements();
+          std::vector<T> pos(num_elements);
+          std::vector<T> len(num_elements);
+          for (size_t i = 0; i < num_elements; ++i) {
+            pos[i] = tensorflow::internal::SubtleMustCopy(pos_flat(i));
+            len[i] = tensorflow::internal::SubtleMustCopy(len_flat(i));
+          }
+          for (size_t i = 0; i < input_tensor.NumElements(); ++i) {
+            // Make sure pos won't cause a runtime error
+            OP_REQUIRES(context, pos[i] >= 0 && pos[i] < input(i).size(),
```

Comment:
Should I be using [`FastBoundsCheck`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/bounds_check.h#L30) here instead of the boolean expression I have?